### DISCUSSION TITLE
How to toggle the icon of a button?

### DISCUSSION THREAD

**Discussion by WolfgangFahl** [zauberzeug/nicegui#1247] [Q&A] [answered]
### Question

http://nicescad.bitplan.com/  has two buttons that would benefit from being able to toggle the icon of the button:
see https://github.com/WolfgangFahl/nicescad/issues/21

* https://nicegui.io/documentation/icon
show how to initialize the icon in the constructor
* https://fonts.google.com/icons
has possible icons
the syntax highlighter should e.g. toggle from "html" to "code"
![grafik](https://github.com/zauberzeug/nicegui/assets/1336221/b1bf49a4-2217-4939-9dbe-bf4642e2c94d)

and  the grid button from "grid on" to "grid off"
![grafik](https://github.com/zauberzeug/nicegui/assets/1336221/e764a893-15cb-4613-bc3c-c46cff9c7886)
![grafik](https://github.com/zauberzeug/nicegui/assets/1336221/5656c9a6-acb1-48de-aebc-49e78d0e8698)

Here is one of my icon toggle attempts:

```python
async def highlight_code(self,_cargs):
        """
        highlight the code and show the html 
        """
        try:
            if self.code_area.visible:
                self.code_area.visible=False
                code_html=self.oscad.highlight_code(self.code)
                self.html_view.content=code_html
                self.html_view.visible=True
                self.highlight_button._props["icon"]="code"
            else:
                self.html_view.visible=False
                self.code_area.visible=True
                self.highlight_button._props["icon"]="html"
            self.highlight_button.update()    
        except BaseException as ex:
            self.handle_exception(ex, self.do_trace)
```
On http://nicescad.bitplan.com/ it seems to work. In my local mac environment it seems to be unreliable.

**Comment by natankeddem** [comment]
Not sure if this is what you are experiencing but I found that on initial boot of the UI, typically rapid firing changes and starting/stopping. I find that ui elements are generally unresponsive and I need to wait a few seconds for things to "settle down". I think the websocket client/server handshaking or something similar is still negotiating.

**Comment by rodja** [comment]
@WolfgangFahl we often archive this by having two buttons side by side and turning one visible and the other invisible. Because the update happens in one sweep, it looks to the user as if the button is toggled.

**Comment by WolfgangFahl** [comment, marked as answer]
@rodja thanks - that's an option i considered. Find my solution below:

```python

 def tool_button(self,tooltip:str,icon:str,handler:callable=None,toggle_icon:str=None)->ui.button:
        """
        Creates an  button with icon that triggers a specified function upon being clicked.
    
        Args:
            tooltip (str): The tooltip to be displayed.
            icon (str): The name of the icon to be displayed on the button.
            handler (function): The function to be called when the button is clicked.
            toggle_icon (str): The name of an alternative icon to be displayed when the button is clicked.
    
        Returns:
            ui.button: The icon button object.
            
        valid icons may be found at:    
            https://fonts.google.com/icons
        """
        icon_button=ui.button("",icon=icon, color='primary').tooltip(tooltip).on("click",handler=handler)  
        icon_button.toggle_icon=toggle_icon
        return icon_button   

 def toggle_icon(self,button:ui.button):
        """
        toggle the icon of the given button
        
        Args:
            ui.button: the button that needs the icon to be toggled
        """
        if hasattr(button,"toggle_icon"):
            # exchange icon with toggle icon
            toggle_icon=button._props["icon"]
            icon=button.toggle_icon
            button._props["icon"]=icon
            button.toggle_icon=toggle_icon
        button.update()
```
